app/src/main/java/org/schabi/newpipe/player/PlayerService.kt

```diff
@@ -0,0 +1,213 @@
+            notification cancelled in its destruction
+             */
+            stopSelf()
+            return START_NOT_STICKY
+        }
+
+        player.handleIntent(intent)
+
+        return START_NOT_STICKY
+    }
+
+    fun stopForImmediateReusing() {
+        if (DEBUG) {
+            Log.d(TAG, "stopForImmediateReusing() called")
+        }
+
+        if (!player.exoPlayerIsNull()) {
+            // Releases wifi & cpu, disables keepScreenOn, etc.
+            // We can't just pause the player here because it will make transition
+            // from one stream to a new stream not smooth
+            player.smoothStopForImmediateReusing()
+        }
+    }
+
+    override fun onTaskRemoved(rootIntent: Intent) {
+        super.onTaskRemoved(rootIntent)
+        if (!player.videoPlayerSelected()) {
+            return
+        }
+        onDestroy()
+        // Unload from memory completely
+        Runtime.getRuntime().halt(0)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        if (DEBUG) {
+            Log.d(TAG, "destroy() called")
+        }
+
+        cleanup()
+
+        mediaBrowserConnector?.release()
+        mediaBrowserConnector = null
+
+        compositeDisposableLoadChildren.clear()
+    }
+
+    private fun cleanup() {
+        player.destroy()
+    }
+
+    fun stopService() {
+        cleanup()
+        stopSelf()
+    }
+
+    override fun attachBaseContext(base: Context) {
+        super.attachBaseContext(AudioServiceLeakFix.preventLeakOf(base))
+    }
+
+    override fun onBind(intent: Intent): IBinder? {
+        if (SERVICE_INTERFACE == intent.action) {
+            return super.onBind(intent)
+        }
+        return mBinder
+    }
+
+    // MediaBrowserServiceCompat methods
+    override fun onGetRoot(
+        clientPackageName: String,
+        clientUid: Int,
+        rootHints: Bundle?,
+    ): BrowserRoot? = mediaBrowserConnector?.onGetRoot(clientPackageName, clientUid, rootHints)
+
+    override fun onLoadChildren(
+        parentId: String,
+        result: Result<List<MediaBrowserCompat.MediaItem>>,
+    ) {
+        result.detach()
+        mediaBrowserConnector?.let {
+            val disposable =
+                it.onLoadChildren(parentId).subscribe { mediaItems ->
+                    result.sendResult(mediaItems)
+                }
+            compositeDisposableLoadChildren.add(disposable)
+        }
+    }
+
+    override fun onSearch(
+        query: String,
+        extras: Bundle,
+        result: Result<List<MediaBrowserCompat.MediaItem>>,
+    ) {
+        mediaBrowserConnector?.onSearch(query, result)
+    }
+
+    class LocalBinder internal constructor(
+        playerService: PlayerService,
+    ) : Binder() {
+        private val playerService = WeakReference(playerService)
+
+        val service: PlayerService?
+            get() = playerService.get()
+
+        fun getPlayer(): Player = service?.player ?: throw Error("Player service is null")
```

Comment:
Is there a reason why a simple `Error` is thrown in this case?

Reply:
I suppose NullPointerException would be more appropriate? Or returning null?

Reply:
You're right; this should be nullable based on its usage. Thanks for catching that.